### PR TITLE
Only put tensors on a device

### PR DESCRIPTION
**Pull request by sgugger**
Fix Trainer when users have inputs containing non-tensor values.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -573,8 +573,7 @@ def _training_step(
         self, model: nn.Module, inputs: Dict[str, torch.Tensor], optimizer: torch.optim.Optimizer
     ) -> float:
         model.train()
-        for k, v in inputs.items():
-            inputs[k] = v.to(self.args.device)
+        inputs = {k: v.to(self.args.device) if isinstance(v, torch.Tensor) else v for k, v in inputs.items()}
```
Comment (julien-c) [inline src/transformers/trainer.py:577]: I find this personally hard to ready.

Maybe:

```python
for k, v in inputs.items():
    if isinstance(v, torch.Tensor):
        inputs[k] = v.to(self.args.device)
```

Comment (julien-c) [reply src/transformers/trainer.py:577]: what do you think?

Comment (julien-c) [reply src/transformers/trainer.py:577]: (otherwise, LGTM – we might want to document this in the function's type signature – or not)

Comment (sgugger) [reply src/transformers/trainer.py:577]: Usually using list-comprehensions are faster, but the batch don't have a lot of keys so it probably doesn't matter.

**Review by julien-c — APPROVED** [review]

**Pull request merged**